### ISSUE TITLE
Use the QEMU Extended Key Event message to (optionally) transmit keycodes instead of keysyms

### ISSUE THREAD

**Issue by dcommander** [TurboVNC/turbovnc#108]
https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#qemu-extended-key-event-message

Should be fairly straightforward but will require a lot of testing.  The main benefits of this:
* The keysyms/key events generated on the server would depend solely on the server's keyboard layout, not the client's.
* Would bypass all of the complex logic that the TurboVNC Viewer and Server currently employ to translate client-side key events into server-side key events.  Because we can't currently transmit keycodes, the viewer has to translate them into keysyms, which depend on the client's keyboard layout, then the server has to figure out how to generate a fake key sequence that will generate the same keysyms using the server's keyboard layout.  This generally works, but it's far from perfect, and it probably doesn't cover all keyboard layout scenarios (TurboVNC has only been tested extensively with a few specific types of international keyboards.)
* It would provide a consistent user experience across multiple client O/S platforms.  If the user presses a particular key, that key would always generate the same X events on the server.  Currently, TurboVNC attempts to produce the same end result for a particular key sequence on the server that would be produced by the key sequence on the client.  For instance, it is expected that pressing a key that generates a "+" on the client will also generate a "+" on the server, but the actual X event sequence used to produce that keysym on the server will vary depending on the client's and the server's keyboard layout.  Games, in particular, tend to rely on raw keycodes rather than keysyms, so they would benefit from this proposed new feature.

**Comment by dcommander** [comment]
Pushed to TurboVNC 2.3

**Comment by dcommander** [comment]
Unfortunately, I ran into a roadblock with this feature after doing some initial investigative work.  Although the Swing API documentation claims that `KeyEvent.getKeyCode()` does not depend on the keyboard layout, with respect to dead keys, it unfortunately does.  A prime example is the dead acute accent key on U.S. International and Brazilian keyboards.  The former is located in the same place as the quote key on U.S. keyboards.  The latter is located in the same place as the left bracket key on U.S. keyboards.  Unfortunately, identical KeyEvents are generated in both cases, which makes it impossible for us to figure out which physical key was pressed (and we need to know that in order to send a keycode to the server rather than a keysym.)

**Comment by dcommander** [comment]
Was able to find a way around the roadblock for Windows at least (fortunately, Java embeds the raw scan code in the parameter string for the KeyEvent.)

**Comment by dcommander** [comment]
This feature has proven more complicated than expected.  Will definitely need funding in order to fully implement and test it.  However, I think that some of what I learned in my research can be applied to work around the issues described in #194.

**Comment by dcommander** [comment]
Implemented in the dev/3.1 evolving [pre-release build](https://turbovnc.org/DeveloperInfo/PreReleases).  Please test thoroughly.  (You need to use both the server and viewer from the pre-release build in order to activate the feature.)

**Comment by davispuh** [comment]
Hmm, I just tried with v3.0 it works correctly regarding #221, I can type accent characters fine but yeah copying clipboard still doesn't work.

But when I try dev build (https://s3.amazonaws.com/turbovnc-pr/dev/linux/turbovnc-3.0.80.tar.gz) then accents doesn't work anymore, when I press dead key, I get the regular character. Copying clipboard works correctly and I can copy/paste/cut Unicode characters.

```
$ vncserver -version
...
TurboVNC Server (Xvnc) 64-bit v3.0.80 (build 20220913)
...
$ java -version
openjdk version "18.0.2" 2022-07-19
OpenJDK Runtime Environment (build 18.0.2+0)
OpenJDK 64-Bit Server VM (build 18.0.2+0, mixed mode)
```

**Comment by davispuh** [comment]
Here's `xev` output normally without VNC when I press dead key \` and then `a` which correctly produces `ā`

```
KeyRelease event, serial 40, synthetic NO, window 0x5e00001,
    root 0x6ab, subw 0x0, time 16976430, (349,169), root:(1220,638),
    state 0x90, keycode 49 (keysym 0x60, grave), same_screen YES,
    XLookupString gives 1 bytes: (60) "`"
    XFilterEvent returns: False

KeyPress event, serial 40, synthetic NO, window 0x5e00001,
    root 0x6ab, subw 0x0, time 16976619, (349,169), root:(1220,638),
    state 0x90, keycode 38 (keysym 0x3e0, amacron), same_screen YES,
    XLookupString gives 2 bytes: (c4 81) "ā"
    XmbLookupString gives 2 bytes: (c4 81) "ā"
    XFilterEvent returns: False

KeyRelease event, serial 40, synthetic NO, window 0x5e00001,
    root 0x6ab, subw 0x0, time 16976690, (349,169), root:(1220,638),
    state 0x10, keycode 38 (keysym 0x61, a), same_screen YES,
    XLookupString gives 1 bytes: (61) "a"
    XFilterEvent returns: False
```


And here's when I do same inside vncviewer
```
KeyPress event, serial 40, synthetic NO, window 0x6000001,
    root 0x535, subw 0x0, time 17479366, (293,602), root:(1182,676),
    state 0x10, keycode 49 (keysym 0x60, grave), same_screen YES,
    XLookupString gives 1 bytes: (60) "`"
    XmbLookupString gives 1 bytes: (60) "`"
    XFilterEvent returns: False

KeyRelease event, serial 40, synthetic NO, window 0x6000001,
    root 0x535, subw 0x0, time 17479466, (293,602), root:(1182,676),
    state 0x10, keycode 49 (keysym 0x60, grave), same_screen YES,
    XLookupString gives 1 bytes: (60) "`"
    XFilterEvent returns: False

KeyPress event, serial 40, synthetic NO, window 0x6000001,
    root 0x535, subw 0x0, time 17479863, (293,602), root:(1182,676),
    state 0x10, keycode 38 (keysym 0x61, a), same_screen YES,
    XLookupString gives 1 bytes: (61) "a"
    XmbLookupString gives 1 bytes: (61) "a"
    XFilterEvent returns: False

KeyRelease event, serial 40, synthetic NO, window 0x6000001,
    root 0x535, subw 0x0, time 17479977, (293,602), root:(1182,676),
    state 0x10, keycode 38 (keysym 0x61, a), same_screen YES,
    XLookupString gives 1 bytes: (61) "a"
    XFilterEvent returns: False
```

**Comment by davispuh** [comment]
When using dev build of vncviewer with x11vnc then it somewhat works
```
KeyPress event, serial 40, synthetic NO, window 0x7e00001,
    root 0x6d0, subw 0x0, time 19163663, (261,79), root:(1390,524),
    state 0x0, keycode 92 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XLookupString gives 0 bytes:
    XmbLookupString gives 0 bytes:
    XFilterEvent returns: False

KeyPress event, serial 40, synthetic NO, window 0x7e00001,
    root 0x6d0, subw 0x0, time 19163664, (261,79), root:(1390,524),
    state 0x80, keycode 38 (keysym 0x3e0, amacron), same_screen YES,
    XLookupString gives 2 bytes: (c4 81) "?"                         // ā
    XmbLookupString gives 2 bytes: (c4 81) "?"                       // ā
    XFilterEvent returns: False

KeyRelease event, serial 40, synthetic NO, window 0x7e00001,
    root 0x6d0, subw 0x0, time 19163664, (261,79), root:(1390,524),
    state 0x80, keycode 92 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XLookupString gives 0 bytes:
    XFilterEvent returns: False
```

but sometimes dead key state gets stuck and outputs accented character even when it shouldn't. Copying Unicode from x11vnc doesn't work.

**Comment by dcommander** [comment]
Thanks for testing the extended clipboard feature.

As for server-side key mapping ... As the name implies, the server-side keyboard layout (the keyboard layout that is active in the TurboVNC session) matters now, and the client-side keyboard layout doesn't.  The TurboVNC Server build system defaults to using the "us" keyboard layout, so you will need to do one of the following in order to change that:

- If you are building TurboVNC from source, you can change the `XKB_DFLT_LAYOUT` and `XKB_DFLT_VARIANT` CMake variables to change the default keyboard layout and variant for your build.
- You can also pass `-xkblayout {layout}` and, if needed, `-xkbvariant {variant}` to `/opt/TurboVNC/bin/vncserver` (or add those arguments to `$serverArgs` in **~/.vnc/turbovncserver.conf** or **/etc/turbovncserver.conf**) to change the keyboard layout and variant at run time.
- You can also change the keyboard layout in your window manager's GUI (for example, in **System-->Preferences-->Hardware-->Keyboard-->Layouts** if you are using MATE.)  This setting should persist across multiple TurboVNC sessions, but some window managers maintain a different set of preferences for TurboVNC and the "root" X server (Display :0).  Thus, this setting won't always be inherited from a local session of the same window manager.
- You can also use `setxkbmap` to temporarily change the keyboard layout for a particular terminal.

I assume that you are using the Latvian (apostrophe) layout.  I verified that I can properly generate a dead macron with the apostrophe key after I invoke `setxkbmap -layout lv -variant apostrophe` from a terminal, and the `xev` output is identical to what you posted above.

Also note: you need to use the 3.1 evolving build of both the TurboVNC Server and the TurboVNC Viewer in order to use server-side key mapping.

**Comment by davispuh** [comment]
That's really strange because I thought I have correct keymap on server
```
$ localectl status
System Locale: LANG=lv_LV.UTF-8
VC Keymap: lv-tilde
X11 Layout: lv
X11 Model: pc105
X11 Variant: tilde
```

but actually indeed it was `us`
```
$ setxkbmap -print
xkb_keymap {
xkb_keycodes  { include "evdev+aliases(qwerty)" };
xkb_types     { include "complete"      };
xkb_compat    { include "complete"      };
xkb_symbols   { include "pc+us+inet(evdev)"     };
xkb_geometry  { include "pc(pc105)"     };
};
```

I didn't know these could go out of sync and I think that shouldn't happen.
I'm actually using `tilde` variant and yeah it works correctly if I update with `setxkbmap`

I'm using Plasma desktop and in keyboard layout it shows same as systemd
![attels](https://user-images.githubusercontent.com/651800/189792604-42d0c7a5-35ad-4b10-9be6-9fd9b03fb296.png)

but if I change it there to different layout then it doesn't get updated neither in `localectl` nor in `setxkbmap` so effectively that setting does nothing and I can have 3 different layouts at same time: `localectl`,  `setxkbmap`  and Plasma.

I think if layout is not specified for vncserver it should use one provided by systemd then I wouldn't have to do anything and it all  would just work.

And yeah I'm testing with both viewer and server build.

**Comment by dcommander** [comment]
> I think if layout is not specified for vncserver it should use one provided by systemd then I wouldn't have to do anything and it all would just work.

OK, now tell me how exactly I'm supposed to do that.

**Comment by davispuh** [comment]
I thought systemd does something more smarter but it just writes/reads `/etc/X11/xorg.conf.d/00-keyboard.conf`, see https://github.com/systemd/systemd/blob/main/src/locale/localed-util.c#L274 and it also can be accessed by [D-Bus](https://www.freedesktop.org/software/systemd/man/org.freedesktop.locale1.html)

So the issue is that real `Xorg` loads config from `/etc/X11/xorg.conf.d/` by default
```
[    30.181] (==) Using config directory: "/etc/X11/xorg.conf.d"
[    30.182] (==) Using system config directory "/usr/share/X11/xorg.conf.d"
```

But `Xvnc` doesn't so if it did read that file then it would use same layout as regular Xorg.

**Comment by dcommander** [comment]
I'll do some testing and try to figure out what's up.

**Comment by dcommander** [comment]
So here's the deal:

The X.org code that reads configuration files (**xorg.conf** and **xorg.conf.d/***) is only present in the "xfree86" device layer, which interfaces the X server with graphics drivers and physical displays.  Virtual-frame-buffer X servers, such as Xvfb and Xvnc, don't (can't, in fact) use that device layer, so they have no inherent ability to read X.org config files.

Apparently **00-keyboard.conf** is a sort of pseudo-standard way for systemd and X.org (XKB, more specifically) to exchange information about the keyboard layout, but they didn't invite any of us remote display software developers to the party.  (They never do, which is why I have spent untold hours over the past ten years working around poor design decisions on the part of systemd developers.)  Personally, I find that mechanism to be sloppy and error-prone.  `localectl` basically gives special treatment to the physical X server by writing a special X.org config file, but nothing prevents another X.org config file from overriding that.  If you put keyboard layout directives into **/etc/X11/xorg.conf.d/01-keyboard.conf**, for instance, then X.org and systemd would no longer agree on the keyboard layout.  If you invoked `localectl` with `set-keymap` instead of `set-x11-keymap`, then X.org and systemd would also no longer agree on the keyboard layout.  They only agree because of a fragile dirty hack.

Because TurboVNC can't use X.org config files, it provides its own equivalent mechanism for setting the default XKB keyboard layout on a per-user or a system-wide basis.  `localectl` won't affect TurboVNC, and I don't see any straightforward fool-proof way of changing that.  Also, the keyboard layout set through `localectl`/systemd is just the pre-login default.  Once you log in, the window manager will override it.  Invoking `localectl` has no immediate effect on the keyboard layout once you are logged in.

In short, changing the keyboard layout through the GUI or `setxkbmap` should work just as well with TurboVNC as it does with a physical X server, but `localectl` is not a mechanism we can currently support.

**Comment by davispuh** [comment]
I completely agree with all your points, especially regrading this `localectl` implementation.

But if you ignore Xorg config files and treat that just as implementation detail, you could use systemd's `org.freedesktop.locale1` D-Bus to get default layout which would be better than compiled in `XKB_DFLT_LAYOUT` since current behavior can be surprising. Anyway it's not really big deal.

**Comment by dcommander** [comment]
The other problems with `localectl` and reading **/etc/X11/xorg.conf.d/00-keyboard.conf**:
1. Only root can set the keyboard layout in that manner, whereas the TurboVNC Server runs under non-root accounts and needs to give those accounts the ability to modify the default XKB layout.
2. It seems as if it's less of a pseudo-standard and more system-specific.  Ubuntu systems, as an example, don't create **/etc/X11/xorg.conf.d/00-keyboard.conf** when you invoke `localectl set-x11-keymap`.  Instead, they modify **/etc/default/keyboard** and set a variable in that file called `XKBLAYOUT`.  I have no idea how that variable is communicated to the X server.

Accessing D-Bus from within the Xvnc process is theoretically possible, but it would be challenging.  Any of the aforementioned mechanisms would have to be tested across the broad range of distributions that we support, which would take a lot of time.  There is no more funding for this feature, so any further modifications to it will need to be straightforward unless another organization steps forward to fund those modifications.

At the end of the day, TurboVNC's primary use case is large-scale on-demand multi-user remote desktop servers.  None of the "regular" users of TurboVNC would have access to run `localectl`, so their primary method of setting the keyboard layout would be through the window manager GUI (which works fine.)

At this point, I'm going to defer the matter until I get feedback from other users, which will undoubtedly happen during the beta test phase.

**Comment by davispuh** [comment]
> Only root can set the keyboard layout in that manner

That's actually not true, you can configure Polkit to allow non-root users invoke this D-Bus, there's 2 parts, the daemon (`systemd`) which is running as root already and your client which doesn't need to run as root.

> Ubuntu systems

they patch it themselves, see https://git.launchpad.net/ubuntu/+source/systemd/commit/src/locale/keymap-util.c?h=applied/ubuntu/jammy-devel&id=35973229b95834155287044d89bc94a7c0f39a7b


>  primary method of setting the keyboard layout would be through the window manager GUI (which works fine.)

for some reason it doesn't work with KDE Plasma, but overall Plasma has been quite buggy regarding VNC so it's possible it's their bug and not TurboVNC issue.

**Comment by dcommander** [comment]
> That's actually not true, you can configure Polkit to allow non-root users invoke this D-Bus, there's 2 parts, the daemon (`systemd`) which is running as root already and your client which doesn't need to run as root.

Right, but non-root cannot modify **/etc/X11/xorg.conf.d/00-keyboard.conf**.  My point was more about how to communicate the keyboard layout to the X server, not how to communicate it to systemd.

> they patch it themselves, see https://git.launchpad.net/ubuntu/+source/systemd/commit/src/locale/keymap-util.c?h=applied/ubuntu/jammy-devel&id=35973229b95834155287044d89bc94a7c0f39a7b

But that doesn't explain how the layout is communicated to the X server.  Mostly, I'm looking for a foolproof way to read the information that is set by `localectl`, and it seems as if D-Bus might be the only way.

> for some reason it doesn't work with KDE Plasma, but overall Plasma has been quite buggy regarding VNC so it's possible it's their bug and not TurboVNC issue.

Yeah, we only claim support for KDE 4 on RHEL/CentOS 7 (https://turbovnc.org/Documentation/Compatibility30), and changing the keyboard layout through the GUI works fine with that version of KDE.  Later versions of RHEL have moved KDE into the EPEL repository, so it isn't part of the base operating system anymore, and it isn't available on Ubuntu unless you install the Kubuntu flavor.  The vast majority of TurboVNC users use GNOME, MATE, or Xfce, so there hasn't been any demand to support KDE 5, and I have never tested KDE 5 personally.

**Comment by dcommander** [comment]
I am happy to fix compatibility issues between TurboVNC and KDE 5, but someone else would need to help me figure out what the issues are and a straightforward approach to fixing them.  I don't have any resources to spend on WMs that are outside of the aforementioned matrix.

**Comment by davispuh** [comment]
> Right, but non-root cannot modify /etc/X11/xorg.conf.d/00-keyboard.conf. My point was more about how to communicate the keyboard layout to the X server, not how to communicate it to systemd.

Well my point was that if you tell `systemd` to change layout, it will write respective config which will be read by X server on next start. So this way it would work for regular user. And as we talked about it already - Xvnc would have to pull it from D-Bus because it can't read config like X org does.

> But that doesn't explain how the layout is communicated to the X server. Mostly, I'm looking for a foolproof way to read the information that is set by localectl, and it seems as if D-Bus might be the only way.

Yeah I have no idea, you would have to go hunt their stuff, maybe they patch X server aswell to read from there... And because Debian/Ubuntu patches `systemd`, you can't rely on any of those files, only on D-Bus. But if they already patch systemd and possibly Xorg, well then they can patch Xvnc aswell and it's not your issue :D

**Comment by davispuh** [comment]
> I am happy to fix compatibility issues between TurboVNC and KDE 5, but someone else would need to help me figure out what the issues are and a straightforward approach to fixing them. I don't have any resources to spend on WMs that are outside of the aforementioned matrix.

I actually might fix it myself at some point, but I already have very long TODO list and this needs quite a bit if investigation to really understand what's happening so fix is probably easiest part once you understand what's going on.
